### PR TITLE
InflightReadsLimiter - limit the memory used by reads end-to-end (from storage/cache to the write to the consumer channel)

### PR DESCRIPTION
**Pull request by eolivelli**
### Motivation

Broker can go out of memory due to many reads enqueued on the PersistentDispatcherMultipleConsumers dispatchMessagesThread (that is used in case of dispatcherDispatchMessagesInSubscriptionThread set to true, that is the default value) The limit of the amount of memory retained due to reads MUST take into account also the entries coming from the Cache.

When dispatcherDispatchMessagesInSubscriptionThread is false (the behaviour of Pulsar 2.10) there is some kind of natural (but still unpredictable!!) back pressure mechanism because the thread that receives the entries from BK of the cache dispatches immediately and synchronously the entries to the consumer and releases them

### Modifications

- Add a new component (InflightReadsLimiter) that keeps track of the overall amount of memory retained due to inflight reads.
- Add a new configuration entry managedLedgerMaxReadsInFlightSizeInMB
- The feature is disabled by default
- Add new metrics to track the status of the broker

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository: https://github.com/eolivelli/pulsar/pull/19

### CODE REVIEW

**Comment by codelipenghui** [comment]
> Broker can go out of memory due to many reads enqueued on the PersistentDispatcherMultipleConsumers dispatchMessagesThread

@eolivelli Does it only happen when the broker has many subscriptions? For one subscription, we always trigger the new read entries operation after sending messages to consumers. We only have a single active read entries operation at a time. Is it able to reproduce?

**Comment by eolivelli** [comment]
@codelipenghui you are correct, the problems arise when you have a broker with many subscriptions on the same topic (and many topics).
There is no broker level guardrail at the moment.
With this patch the memory used to handle outbound traffic is capped and the limit is independent from the number of active subscriptions on the broker

**Review by nicoloboschi — APPROVED** [review]
LGTM

**Review by codelipenghui — COMMENTED** [review]
Great change.
I have some minor comments.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/cache/RangeEntryCacheImpl.java [hunk 1]
```diff
@@ -313,6 +349,74 @@ void asyncReadEntry0(ReadHandle lh, long firstEntry, long lastEntry, boolean sho
...
+            });
+            return null;
+        } else {
+            callback = new AsyncCallbacks.ReadEntriesCallback() {
+
+                @Override
+                public void readEntriesComplete(List<Entry> entries, Object ctx) {
+                    if (!entries.isEmpty()) {
+                        long size = entries.get(0).getLength();
+                        estimatedEntrySize = size;
```
[the remark this answers is not in the document]
Comment (eolivelli) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/cache/RangeEntryCacheImpl.java:394]: Unfortunately here we are in "managed-ledger" module and in order to get a value from the Dispatcher/Consumer I would have to change many internal APIs.

If the size of entries in the topics is similar for all the entries that I think that this is a good estimate.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/EntryImpl.java [hunk 2]
```diff
@@ -44,6 +44,8 @@ protected EntryImpl newObject(Handle<EntryImpl> handle) {
     private long entryId;
     ByteBuf data;
 
+    private Runnable onDellocate;
```
[the remark this answers is not in the document]
Comment (eolivelli) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/EntryImpl.java:47]: good catch !
committed

**Review by codelipenghui — APPROVED** [review]

**Review by lhotari — APPROVED** [review]

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/EntryImpl.java [hunk 3]
```diff
@@ -44,6 +44,8 @@ protected EntryImpl newObject(Handle<EntryImpl> handle) {
     private long entryId;
     ByteBuf data;
 
+    private Runnable onDeallocate;
```
Comment (Jason918) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/EntryImpl.java:47]: This is not volatile, but it seems that `deallocate()` could be called in other threads?

Comment (eolivelli) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/EntryImpl.java:47]: this is the pattern we use for every recyclable object, the same comment may apply to all the other fields.

**Review by Jason918 — APPROVED** [review]

**Pull request merged**